### DISCUSSION TITLE
Too many template parameters template_arg_list_different_arity

### DISCUSSION THREAD

**Discussion by jobinnthomas** [clangd/clangd#1899] [Q&A]
//------------------------------------------------------------------------------------------
// This struct provides case-insensitive comparison of std::map or std::set keys
// when used like this
//		std::map<CComBSTR, anyType, CComBSTRCmpIgnoringCase>
//------------------------------------------------------------------------------------------
struct CComBSTRCmpIgnoringCase : std::binary_function<CComBSTR, CComBSTR, bool>
{
	bool operator()(const CComBSTR & sLeft, const CComBSTR & sRight) const
	{
		CLecComBSTR temp(sLeft);
		return (temp.CmpIgnoringCase(sRight) < 0);
	}
};
Using the above code, Microsoft visual C++ v2015 does not show any errors but Clang v17.0.3 has a compile error 'template_arg_list_different_arity'   . I saw this thread https://youtrack.jetbrains.com/issue/CPP-25855/Incorrect-Clangd-error-for-partial-template-specialization-with-default-parameters  and tried to use the below configuration in my config.yaml but I am still getting the error. Is this the right way to do this?

CompileFlags:
    Add:         
        - -Wno-non-pod-varargs
        - -I"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\atlmfc\\include"
        - -I"H:\\mauisoft\\main\\shared\\includes"        
        - -fno-relaxed-template-template-args        
       
Diagnostics:
    Suppress:
        - -Wwritable-strings
        - -Wtautological-pointer-compare

**Comment by HighCommander4** [comment]
It's hard to say anything without a complete code example and complete error output.

**Reply by jobinnthomas** [reply]
Here is a code example that fails with the same error. I tried to add '-fno-relaxed-template-template-args' using the below configuration in my config.yaml but I am still getting the error. Is there a way to ignore this error? i cannot change the code right now.

CompileFlags:
Add:
- -fno-relaxed-template-template-args

```
#include <iostream>

template <template <typename, typename> class Op>
class Function
{
};

template <typename A, typename B, bool is_f = std::is_floating_point<A>::value || std::is_floating_point<B>::value > struct Operator;

template <typename A, typename B>
struct Operator<A, B, false>
{};


template <typename A, typename B>
struct Operator<A, B, true>
{};

using FunctionOperator = Function<Operator>;


int main(int argc, char * argv[]){
    std::cout << "hi!\n";
    return 0;
}
```
Compile errors:

tmp.cpp:19:35: error: template template argument has different template parameters than its corresponding template template parameter
using FunctionOperator = Function<Operator>;
                                  ^
tmp.cpp:8:1: note: too many template parameters in template template argument
template <typename A, typename B, bool is_f = std::is_floating_point<A>::value || std::is_floating_point<B>::value > struct Operator;
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tmp.cpp:3:11: note: previous template template parameter is here
template <template <typename, typename> class Op>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

**Reply by HighCommander4** [reply]
Thanks, I see now.

It looks like this concerns a C++17 language feature called "Matching template template parameters to compatible arguments". Its entry in the table at https://clang.llvm.org/cxx_status.html#p0522 is marked as "Partial", with the following comment:

> Despite being the resolution to a Defect Report, this feature is disabled by default in all language versions, and can be enabled explicitly with the flag `-frelaxed-template-template-args` in Clang 4 onwards.

So, you almost had it right: you need to **enable** the "relaxed template template args" flag with`-frelaxed-template-template-args` rather than disabling it with `-fno-relaxed-template-template-args`.
